salt/utils/win_dacl.py

```diff
@@ -1115,34 +1121,37 @@ def get_name(principal):
         salt.utils.win_dacl.get_name('S-1-5-32-544')
         salt.utils.win_dacl.get_name('adminisTrators')
     '''
-    # If None is passed, use the Universal Well-known SID for "Null SID"
-    if principal is None:
-        principal = 'S-1-0-0'
-
-    # Assume PySID object
-    sid_obj = principal
-
-    # Name
-    try:
-        sid_obj = win32security.LookupAccountName(None, principal)[0]
-    except (pywintypes.error, TypeError):
-        pass
-
-    # String SID
-    try:
-        sid_obj = win32security.ConvertStringSidToSid(principal)
-    except (pywintypes.error, TypeError):
-        pass
+    # If this is a PySID object, use it
+    if isinstance(principal, pywintypes.SIDType):
+        sid_obj = principal
+    else:
+        # If None is passed, use the Universal Well-known SID for "Null SID"
+        if principal is None:
+            principal = 'S-1-0-0'
+
+        # String SID
+        # Try Converting String SID to SID Object first as it's least expensive
+        try:
+            sid_obj = win32security.ConvertStringSidToSid(principal)
+        except pywintypes.error:
+
+            # Name
+            # Try Getting the SID Object by Name Lookup last
+            # This is expensive, especially on large AD Domains
+            try:
+                sid_obj = win32security.LookupAccountName(None, principal)[0]
+            except pywintypes.error:
+                # Just pass the object and let the LookupAccountSid function try
+                sid_obj = principal
 
+    # By now we should have a valid PySID object
     try:
         return win32security.LookupAccountSid(None, sid_obj)[0]
     except (pywintypes.error, TypeError) as exc:
+        message = 'Error resolving "{0}"'.format(principal)
         if type(exc) == pywintypes.error:
             win_error = win32api.FormatMessage(exc.winerror).rstrip('\n')
```

Comment:
Shouldn't this be `message =` instead of `win_error =` (since `win_error` is not used after this line). If so, then you can move the `message` that is currently above the `if` into the `else` (unless you want to use it in this `if` case in addition to the message from `FormatMessage`).

Reply:
It's used on the following line to add more detail to the error message if it is a `pywintypes.error`

Reply:
I see what you are saying. I got confused because the `else` case was red (removed), not green (added). So there is no `else` case at all anymore. Ignore that comment.